src/parse/parse_mat.jl

```diff
@@ -0,0 +1,1667 @@
+# CONTRIB file
+# Originally from PowerModels.jl. Adapted by PowerSystems.jl and by us.
+
+#########################################################################
+#                                                                       #
+# This file provides functions for interfacing with Matpower data files #
+#                                                                       #
+#########################################################################
+
+module ParseMAT
+
+using ..ExaPF: IdxSet
+
+include("matlab.jl")
+include("im_common.jl")
+
+"Parses the matpwer data from either a filename or an IO object"
+function parse_matpower(io::IO)::Dict
+    mp_data = _parse_matpower_string(read(io, String))
+    #pm_data = _matpower_to_powermodels!(mp_data)
+    return mp_data
+end
+
+function parse_mat(file::String; kwargs...)::Dict
+    mp_data = open(file) do io
+        parse_matpower(io; kwargs...)
+    end
+    return mp_data
+end
+
+function mat_to_exapf(data_mat)
+    
+    # create new data structure
+    data = Dict{String,Array}()
+    
+    # baseMVA
+    data["baseMVA"] = [data_mat["baseMVA"]]
+
+    # buses
+    BUS_I, BUS_TYPE, PD, QD, GS, BS, BUS_AREA, VM, VA, BASE_KV, ZONE, VMAX, VMIN,
+    LAM_P, LAM_Q, MU_VMAX, MU_VMIN = IdxSet.idx_bus()
+    
+    nbus = length(data_mat["bus"])
+    bus_array = zeros(nbus, 17)
+
+    for (i, bus) in enumerate(data_mat["bus"])
+        busn = i # we need an external2internal function
+        bus_array[busn, BUS_I] = busn
+        bus_array[busn, BUS_TYPE] = bus["bus_type"]
+        bus_array[busn, PD] = bus["pd"]
+        bus_array[busn, PD] = bus["pd"]
+        bus_array[busn, QD] = bus["qd"]
+        bus_array[busn, GS] = bus["gs"]
+        bus_array[busn, BS] = bus["bs"]
+        bus_array[busn, BUS_AREA] = bus["area"]
+        bus_array[busn, VM] = bus["vm"]
+        bus_array[busn, VA] = bus["va"]
+        bus_array[busn, BASE_KV] = bus["base_kv"]
+        bus_array[busn, ZONE] = bus["zone"]
+        bus_array[busn, VMAX] = bus["vmax"]
+        bus_array[busn, VMIN] = bus["vmin"]
+    end
+    data["bus"] = bus_array
+
+    # generators
+    GEN_BUS, PG, QG, QMAX, VG, MBASE, GEN_STATUS, PMAX, PMIN, PC1, PC2, QC1MIN,
+    QC2MIN, QC2MAX, RAMP_AGC, RAMP_10, RAMP_30, RAMP_Q, APF, MU_PMAG, MU_PMIN, MU_QMAX,
+    MU_QMIN = IdxSet.idx_gen()
+    
+    ngen = length(data_mat["gen"])
+    gen_array = zeros(ngen, 25)
+
+    for (i, gen) in enumerate(data_mat["gen"])
+        gen_array[i, GEN_BUS] = gen["gen_bus"]
+        gen_array[i, PG] = gen["pg"]
+        gen_array[i, QG] = gen["qg"]
+        gen_array[i, QMAX] = gen["qmax"]
+        gen_array[i, VG] = gen["vg"]
+        gen_array[i, MBASE] = gen["mbase"]
+        gen_array[i, GEN_STATUS] = gen["gen_status"]
+        gen_array[i, PMAX] = gen["pmax"]
+        gen_array[i, PMIN] = gen["pmin"]
+        gen_array[i, PC1] = gen["pc1"]
```

Comment:
Here we could have a small problem. It appears that not all matpower instances have fields `pc1`, `pc2`, `qc1min` and `qc2min`. I do not know if we need this information afterwards. But I think at some point we should add a test to check if the dictionary `gen` has these keys (we could use the function `haskey` for instance)